certora/funnels/DepositorUniV3.spec

```diff
@@ -0,0 +1,696 @@
+    assert endOtherAfter == endOtherBefore, "withdraw did not keep unchanged the rest of limits[x][y][z].end";
+    assert gem0BalanceOfBufferAfter == gem0BalanceOfBufferBefore + col0, "withdraw did not increase gem0.balanceOf(buffer) by col0";
+    assert gem1BalanceOfBufferAfter == gem1BalanceOfBufferBefore + col1, "withdraw did not increase gem1.balanceOf(buffer) by col1";
+    assert gem0BalanceOfPoolAfter == gem0BalanceOfPoolBefore - col0, "withdraw did not decrease gem0.balanceOf(pool) by col0";
+    assert gem1BalanceOfPoolAfter == gem1BalanceOfPoolBefore - col1, "withdraw did not decrease gem1.balanceOf(pool) by col1";
+    assert retLiq == liquidity, "withdraw did not return the expected liquidity";
+    assert retAmt0 == amt0, "withdraw did not return the expected amt0";
+    assert retAmt1 == amt1, "withdraw did not return the expected amt1";
+    assert retFees0 == col0 - amt0, "withdraw did not return the expected col0 - amt0";
+    assert retFees1 == col1 - amt1, "withdraw did not return the expected col1 - amt1";
+}
+
+// Verify revert rules on withdraw
+rule withdraw_revert(DepositorUniV3.LiquidityParams p, bool takeFees) {
+    env e;
+
+    require p.gem0 == gem0Con;
+    require p.gem1 == gem1Con;
+    require p.gem0 == poolCon.gem0();
+    require p.gem1 == poolCon.gem1();
+    require p.fee  == poolCon.fee();
+
+    require poolCon.random2() >= poolCon.random0();
+    require poolCon.random3() >= poolCon.random1();
+
+    require e.block.timestamp <= max_uint32;
+
+    address buffer = buffer();
+    require buffer != currentContract;
+    require buffer != poolCon;
+
+    bool canCall = roles.canCall(ilk(), e.msg.sender, currentContract, to_bytes4(0xcd8e305c));
+    mathint wardsSender = wards(e.msg.sender);
+    mathint cap0Gem0Gem1Fee; mathint cap1Gem0Gem1Fee; mathint eraGem0Gem1Fee; mathint due0Gem0Gem1Fee; mathint due1Gem0Gem1Fee; mathint endGem0Gem1Fee;
+    cap0Gem0Gem1Fee, cap1Gem0Gem1Fee, eraGem0Gem1Fee, due0Gem0Gem1Fee, due1Gem0Gem1Fee, endGem0Gem1Fee = limits(p.gem0, p.gem1, p.fee);
+    mathint amt0 = poolCon.random0();
+    mathint amt1 = poolCon.random1();
+    mathint col0 = takeFees ? poolCon.random2() : amt0;
+    mathint col1 = takeFees ? poolCon.random3() : amt1;
+    mathint gem0BalanceOfPool = gem0Con.balanceOf(poolCon);
+    mathint gem1BalanceOfPool = gem1Con.balanceOf(poolCon);
+    mathint due0Updated = to_mathint(e.block.timestamp) >= endGem0Gem1Fee ? cap0Gem0Gem1Fee : due0Gem0Gem1Fee;
+    mathint due1Updated = to_mathint(e.block.timestamp) >= endGem0Gem1Fee ? cap1Gem0Gem1Fee : due1Gem0Gem1Fee;
+
+    withdraw@withrevert(e, p, takeFees);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = !canCall && wardsSender != 1;
+    bool revert3 = p.gem0 >= p.gem1;
+    bool revert4 = to_mathint(e.block.timestamp) >= endGem0Gem1Fee && e.block.timestamp + eraGem0Gem1Fee > max_uint32;
+    bool revert5 = gem0BalanceOfPool < col0;
+    bool revert6 = gem1BalanceOfPool < col1;
```

Comment:
Should reverts 5 & 6 be moved to a `require` instead? Seems that this relates to the inner logic of UniV3 and so can be assumed to be correct, no?

Reply:
Yeah, agreed and done.